### PR TITLE
Spells: Implement target threshold sqrt aoe reduction

### PR DESCRIPTION
**Pull request by Seyden**
**Changes proposed:**

Added 2 new script methods to overwrite target caps before hit hooks happen

-  Add SpellScript::SetSoftTargetCap
-  Add SpellScript::SetFullAmountTargetCap

SoftTargetCap is an value that determines how many targets can be hit before the square root aoe reduction kicks in. 
Default value is 0.

FullAmountTargetCap when used with SoftTargetCap > 0 determines how many target(s) will take the full unreduced amount. Default value is 0.

### **Important note!!!:**
**The soft target cap does not overwrite the aoe damage target cap, if a spell has a soft target cap listed in wowhead as 20, then its not needed to set it.**

**If the soft target cap is set to 6, it will only scale up to 20 and afterwards will be cut through 20 targets.**

The calculation was taken from Simcraft, they tested the behaviour throughly on the PTR

**Issues addressed:**

Closes #  (insert issue tracker number)

**Tests performed:**

(Does it build, tested in-game, etc.)
- [x] Builds
- [x] Tested in-game

### CODE REVIEW

**Comment by Nyr97** [comment]
I'm not sure the implementation is correct for the soft-cap since it's a known fact that class spells with both damage and healing effects have different soft-caps, which basically affects healing specializations across the board:

Chi Burst: damage soft-cap 20, healing soft-cap 6.

Hammer of Light: damage soft-cap 20, healing soft-cap 6.

Halo: damage soft-cap 20, healing soft-cap 6.

Divine Star: damage soft-cap 20, healing soft-cap 6.

**Comment by Shauren** [comment]
All of the spells you mentioned (well idk about hammer of light) are areatriggers that cast single targeting spells on each individual unit and will not interact with this PR at all

**Comment by Nyr97** [comment]
Light's Hammer summons a npc, the caster gets aura 114918, which triggers 119952 (AoE healing) on the npc.

**Comment by Seyden** [comment]
> Not sure the implementation is correct, BUT it's a known fact that class spells with both damage and healing effects have different soft-caps, which basically affects healing specializations across the board:
> 
> Chi Burst: damage soft-cap 20, healing soft-cap 6.
> 
> Light's Hammer: damage soft-cap 20, healing soft-cap 6.
> 
> Halo: damage soft-cap 20, healing soft-cap 6.
> 
> Divine Star: damage soft-cap 20, healing soft-cap 6.

I'm not quite sure what your intention is by posting this list, there is no question or anything inside your comment.

If your spell has a soft cap, you can assign a soft cap to it.

If your healing spell has a soft cap of 6, then you can set it to 6.

You dont have to set the soft cap for 20 targets anyway because its per default since wotlk for all AOE spells.

**Comment by Nyr97** [comment]
> Light's Hammer summons a NPC, its caster gets aura 114918, which triggers 119952 (AoE healing) periodically on the NPC's location, so yes, the spell will be affected by this.

This isn't supported correctly by your implementation.

**Comment by Seyden** [comment]
> > Light's Hammer summons an NPC, and its caster gets aura 114918, which triggers 119952 (AoE healing) periodically on the NPC's location, so yes, the spell will be affected by this.
> 
> This won't be supported correctly by your implementation because the paladin triggers an AoE healing periodically on the summoned dummy, it's not an individual heal cast by an areatrigger.

I dont get what you mean, every cast which is a AOE, you can set the Soft Target Cap to 6, i dont see the problem?

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Spells/Spell.cpp [hunk 1]
```diff
@@ -8344,10 +8363,18 @@ void Spell::DoEffectOnLaunchTarget(TargetInfo& targetInfo, float multiplier, Spe
...
+                    && m_softTargetCapCount > 0
+                    && targetCount > m_softTargetCapCount)
+                    m_damage *= std::sqrt(float(m_softTargetCapCount) / std::min<int64>(AOE_DAMAGE_TARGET_CAP, targetCount));
+
                 // cap damage of player AOE
-                int64 targetAmount = GetUnitTargetCountForEffect(spellEffectInfo.EffectIndex);
-                if (targetAmount > 20)
-                    m_damage = m_damage * 20 / targetAmount;
+                if (targetCount > AOE_DAMAGE_TARGET_CAP)
+                    m_damage *= AOE_DAMAGE_TARGET_CAP / targetCount;
```
[the remark this answers is not in the document]
Comment (Seyden) [reply src/server/game/Spells/Spell.cpp]: Thank you for noticing !

**Comment by Seyden** [comment]
@Shauren just if you missed the commit comments by any chance..

there are multiple spells which dynamically update the target cap based on talents and stuff..



ID - 198013 Eye Beam
Takes it from EFFECT_5

ID - 258926 Fel Barrage
Takes it from EFFECT_2

ID - 400370 Burning Vehemence
Is also taking it from basepoints and this is also a talent

This is also a talent which modifies another spell
https://www.wowhead.com/spell=403826/blade-of-vengeance

Deep Impact from Mage is also dynamic

**Pull request merged**